File: modules/14-local-inference-optimization/ts/03-throughput-latency.ts
/**
 * Task 3 🟡 — Throughput vs latency: batching and concurrency.
 *
 * What you'll learn:
 *   - The difference between latency (single request) and throughput (many)
 *   - Time-to-first-token (TTFT): the most user-perceived metric in chat apps
 *   - How concurrent requests improve aggregate throughput
 *   - Why batching is fundamental for high-throughput serving
 *
 * Key insight: a single request gets the lowest latency by monopolising resources.
 * But aggregate throughput across all users is maximised by batching requests —
 * the weight-loading overhead is amortised.
 *
 * How to run:
 *   pnpm tsx modules/14-local-inference-optimization/ts/03-throughput-latency.ts
 */

import { getProvider } from "@learn-ai/llm-core";

const PROMPT =
  "List 5 practical tips for writing clean, maintainable TypeScript code. " +
  "Be concise — one sentence per tip.";

const SHORT_PROMPT = "What is 2 + 2? Answer with just the number.";

// ---------------------------------------------------------------------------
// Time-to-first-token (TTFT)
// ---------------------------------------------------------------------------

/**
 * Measure the time from sending a request to receiving the FIRST token.
 *
 * Uses chatStream() to capture the exact moment the first chunk arrives.
 * TTFT is the dominant latency for interactive chat UIs.
 *
 * Returns elapsed milliseconds until the first token.
 *
 * TODO:
 *   1. Snapshot start = performance.now().
 *   2. Iterate provider.chatStream() over a single "user" message with
 *      `for await (const chunk of ...)`.
 *   3. The moment you see the FIRST non-empty chunk, snapshot the clock and
 *      break — don't drain the rest of the stream.
 *   4. Return the elapsed ms from start to that first chunk (the caller converts
 *      to seconds).
 */
async function measureTtft(
  prompt: string,
  provider: ReturnType<typeof getProvider>,
): Promise<number> {
  // TODO: implement measureTtft
  throw new Error("TODO: implement measureTtft()");
}

// ---------------------------------------------------------------------------
// Single-request latency
// ---------------------------------------------------------------------------

interface SingleResult {
  ttftS: number;
  totalS: number;
  tokensOut: number;
  tokensPerS: number;
}

/**
 * Measure total latency, TTFT, and tokens/sec for one request.
 *
 * TODO:
 *   1. Measure TTFT with measureTtft().
 *   2. Measure full request: performance.now() + provider.chat().
 *   3. Return { ttftS, totalS, tokensOut, tokensPerS }.
 */
async function measureSingleRequest(
  prompt: string,
  provider: ReturnType<typeof getProvider>,
): Promise<SingleResult> {
  // TODO: implement measureSingleRequest
  throw new Error("TODO: implement measureSingleRequest()");
}

// ---------------------------------------------------------------------------
// Concurrent requests
// ---------------------------------------------------------------------------

interface ConcurrentResult {
  n: number;
  wallClockS: number;
  totalTokens: number;
  aggregateTokensPerS: number;
}

/**
 * Fire `n` requests concurrently and measure aggregate throughput.
 *
 * Returns { n, wallClockS, totalTokens, aggregateTokensPerS }.
 *
 * TODO:
 *   1. Snapshot start = performance.now().
 *   2. Kick off n provider.chat() calls at once and await them together with
 *      Promise.all — keep each run short with a small maxTokens.
 *   3. Snapshot end; wallClockS is the difference in seconds.
 *   4. Sum the output tokens across all results.
 *   5. Return the ConcurrentResult (n / wallClockS / totalTokens /
 *      aggregateTokensPerS).
 */
async function measureConcurrent(
  prompt: string,
  provider: ReturnType<typeof getProvider>,
  n: number,
): Promise<ConcurrentResult> {
  // TODO: implement measureConcurrent
  throw new Error("TODO: implement measureConcurrent()");
}

// ---------------------------------------------------------------------------
// Latency table
// ---------------------------------------------------------------------------

/**
 * Print a table comparing single-request latency vs concurrent throughput.
 *
 * Columns: Concurrency | Wall-clock(s) | Total tokens | Agg. Tokens/sec
 *
 * TODO: print a header row and one row per result.
 */
function printLatencyTable(results: ConcurrentResult[]): void {
  // TODO: implement printLatencyTable
  throw new Error("TODO: implement printLatencyTable()");
}

// ---------------------------------------------------------------------------
// Harness
// ---------------------------------------------------------------------------

async function main() {
  const provider = getProvider();
  console.log(`Provider: ${provider.name} | Model: ${provider.chatModel}`);

  // Single-request baseline
  console.log("\n" + "=".repeat(60));
  console.log("SINGLE REQUEST BASELINE");
  console.log("=".repeat(60));
  try {
    const single = await measureSingleRequest(PROMPT, provider);
    console.log(`  TTFT           : ${single.ttftS.toFixed(3)}s`);
    console.log(`  Total latency  : ${single.totalS.toFixed(3)}s`);
    console.log(`  Output tokens  : ${single.tokensOut}`);
    console.log(`  Tokens/sec     : ${single.tokensPerS.toFixed(1)}`);
  } catch (e) {
    console.log(`  Skipped: ${e}`);
    return;
  }

  // Concurrent benchmark
  console.log("\n" + "=".repeat(60));
  console.log("CONCURRENT THROUGHPUT (N parallel requests)");
  console.log("=".repeat(60));

  const tableResults: ConcurrentResult[] = [];
  for (const n of [1, 2, 4, 8]) {
    try {
      process.stdout.write(`  n=${n}... `);
      const res = await measureConcurrent(SHORT_PROMPT, provider, n);
      console.log(
        `wall=${res.wallClockS.toFixed(2)}s  agg=${res.aggregateTokensPerS.toFixed(1)} tok/s`,
      );
      tableResults.push(res);
    } catch (e) {
      console.log(`skipped (${e})`);
    }
  }

  console.log();
  printLatencyTable(tableResults);

  console.log(
    "\nKey takeaway: concurrent requests increase aggregate throughput because",
    "\nthe model server can batch work. Wall-clock time grows sub-linearly with N.",
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: modules/14-local-inference-optimization/py/03_throughput_latency.py
"""
Task 3 🟡 — Throughput vs latency: batching and concurrency.

What you'll learn:
  - The difference between latency (single request) and throughput (many requests)
  - Time-to-first-token (TTFT): the most user-perceived metric in chat apps
  - How concurrent requests improve aggregate throughput
  - Why batching is the fundamental mechanism for high-throughput serving

Key insight: a single request gets the lowest latency by monopolising the GPU/CPU.
But throughput (total tokens/sec across all users) is maximised by batching multiple
requests together — the weight-loading overhead is amortised.

How to run:
  uv run python modules/14-local-inference-optimization/py/03_throughput_latency.py
"""

from __future__ import annotations

import asyncio
import time

from llm_core import ChatMessage, ChatOptions, get_provider

PROMPT = (
    "List 5 practical tips for writing clean, maintainable Python code. "
    "Be concise — one sentence per tip."
)

SHORT_PROMPT = "What is 2 + 2? Answer with just the number."


# ---------------------------------------------------------------------------
# Time-to-first-token (TTFT)
# ---------------------------------------------------------------------------


def measure_ttft(prompt: str, provider) -> float:
    """
    Measure the time from sending a request to receiving the FIRST token.

    This uses streaming (chat_stream) to capture the exact moment the first
    chunk arrives. TTFT is the dominant latency for interactive chat UIs.

    Returns elapsed seconds until the first token.

    TODO:
      1. Snapshot start = time.perf_counter().
      2. Iterate provider.chat_stream() over a single "user" message.
      3. The moment you see the FIRST non-empty chunk, snapshot the clock and
         break — don't drain the rest of the stream.
      4. Return the elapsed seconds from start to that first chunk.

    Note: we break immediately after the first chunk to avoid waiting for
    the full response — that's the whole point of TTFT.
    """
    # TODO: implement measure_ttft
    raise NotImplementedError("TODO: implement measure_ttft()")


# ---------------------------------------------------------------------------
# Single-request latency
# ---------------------------------------------------------------------------


def measure_single_request(prompt: str, provider) -> dict:
    """
    Measure total latency, TTFT, and tokens/sec for one request.

    Returns:
      "ttft_s"      : time to first token in seconds
      "total_s"     : total response time in seconds
      "tokens_out"  : number of output tokens
      "tokens_per_s": tokens_out / total_s

    TODO:
      1. Measure TTFT with measure_ttft().
      2. Measure full request with time.perf_counter() + provider.chat().
      3. Return the dict.
    """
    # TODO: implement measure_single_request
    raise NotImplementedError("TODO: implement measure_single_request()")


# ---------------------------------------------------------------------------
# Concurrent requests
# ---------------------------------------------------------------------------


async def _send_request_async(prompt: str, provider) -> dict:
    """Run one chat request in a thread and return timing info."""
    loop = asyncio.get_running_loop()
    start = time.perf_counter()
    result = await loop.run_in_executor(
        None,
        lambda: provider.chat([ChatMessage("user", prompt)], ChatOptions(max_tokens=100)),
    )
    elapsed = time.perf_counter() - start
    tokens_out = result.usage.output_tokens or 0
    return {
        "tokens_out": tokens_out,
        "elapsed_s": elapsed,
        "tokens_per_s": tokens_out / elapsed if elapsed > 0 else 0,
    }


async def measure_concurrent_async(prompt: str, provider, n: int) -> dict:
    """
    Fire `n` requests concurrently and measure aggregate throughput.

    Returns:
      "n"                  : number of concurrent requests
      "wall_clock_s"       : total wall-clock time (from first send to last receive)
      "total_tokens"       : sum of all output tokens
      "aggregate_tokens_per_s": total_tokens / wall_clock_s

    TODO:
      1. Record start = time.perf_counter().
      2. Use asyncio.gather(*[_send_request_async(prompt, provider) for _ in range(n)]).
      3. Record end time.
      4. Compute wall_clock_s = end - start.
      5. Sum tokens from all results.
      6. Return the dict.
    """
    # TODO: implement measure_concurrent_async
    raise NotImplementedError("TODO: implement measure_concurrent_async()")


def measure_concurrent(prompt: str, provider, n: int) -> dict:
    """Sync wrapper for measure_concurrent_async."""
    return asyncio.run(measure_concurrent_async(prompt, provider, n))


# ---------------------------------------------------------------------------
# Latency table
# ---------------------------------------------------------------------------


def print_latency_table(results: list[dict]) -> None:
    """
    Print a table comparing single-request latency vs concurrent throughput.

    Expected columns: Concurrency | Wall-clock(s) | Tokens | Agg. Tokens/sec

    TODO:
      1. Print a header row.
      2. For each result in results, print a formatted row.
         result keys: "n", "wall_clock_s", "total_tokens", "aggregate_tokens_per_s"
         (results from measure_single_request have different keys — adapt accordingly)
    """
    # TODO: implement print_latency_table
    raise NotImplementedError("TODO: implement print_latency_table()")


# ---------------------------------------------------------------------------
# Harness
# ---------------------------------------------------------------------------


def main() -> None:
    provider = get_provider()
    print(f"Provider: {provider.name} | Model: {provider.chat_model}")

    # Single-request baseline
    print("\n" + "=" * 60)
    print("SINGLE REQUEST BASELINE")
    print("=" * 60)
    try:
        single = measure_single_request(PROMPT, provider)
        print(f"  TTFT           : {single['ttft_s']:.3f}s")
        print(f"  Total latency  : {single['total_s']:.3f}s")
        print(f"  Output tokens  : {single['tokens_out']}")
        print(f"  Tokens/sec     : {single['tokens_per_s']:.1f}")
    except Exception as e:
        print(f"  Skipped: {e}")
        return

    # Concurrent benchmark
    print("\n" + "=" * 60)
    print("CONCURRENT THROUGHPUT (N parallel requests)")
    print("=" * 60)

    table_results = []
    for n in [1, 2, 4, 8]:
        try:
            print(f"  n={n}...", end=" ", flush=True)
            res = measure_concurrent(SHORT_PROMPT, provider, n)
            print(f"wall={res['wall_clock_s']:.2f}s  agg={res['aggregate_tokens_per_s']:.1f} tok/s")
            table_results.append(res)
        except Exception as e:
            print(f"skipped ({e})")

    print()
    print_latency_table(table_results)

    print(
        "\nKey takeaway: concurrent requests increase aggregate throughput because"
        "\nthe model (or Ollama server) can batch work. Wall-clock time grows sub-"
        "\nlinearly with N — especially visible at higher concurrency."
    )


if __name__ == "__main__":
    main()
